tests/virt_autotest/vgpu.pm

```diff
@@ -0,0 +1,368 @@
+    record_info("vGPU Grid driver is installed successfully in $vm");
+    # Verify
+    assert_script_run("ssh root\@$vm 'lsmod | grep -i nvidia'");
+}
+
+sub detach_vgpu_from_guest {
+    my ($slot, $vm) = @_;
+    assert_script_run("ssh root\@$vm 'poweroff'");
+    script_retry("virsh domstate $vm | grep 'shut off'", timeout => 60, delay => 5, retry => 3, die => 0);
+    assert_script_run("virt-xml $vm --remove-device --hostdev type=mdev,address.slot=$slot");
+    assert_script_run("virsh start $vm");
+    record_info("vGPU has been removed successfully from $vm");
+    wait_guest_online($vm);
+}
+
+sub remove_vgpu {
+    my $uuid = shift;
+    assert_script_run("cd `find /sys/devices -type d -name $uuid`");
+    assert_script_run("echo '1' > remove");
+    script_run('cd ../mdev_supported_types/; for i in *; do echo "$i" $(cat $i/name) available: $(cat $i/avail*); done');
+    die "Fatal: vGPU device $uuid is still alive!" if script_run("lsmdev | grep $uuid") == 0;
+    record_info("vGPU removed", "$uuid has been removed from the host");
+}
+
+sub save_vgpu_device_status_logs {
+    my $vm = shift;
+
+    #vm configuration file
+    enter_cmd("virsh dumpxml $vm > $log_dir/${vm}.xml");
+
+    my $log_file = "log.txt";
+    #logging device information in guest
+    if (script_run("virsh domstate $vm | grep running") == 0) {
+        script_run "echo `date` > $log_file";
+        script_run "echo '***** Status & logs inside $vm *****' >> $log_file";
+        my $debug_script = "vgpu_guest_logging.sh";
+        download_script($debug_script, machine => $vm, proceed_on_failure => 1);
+        script_run("timeout 25 ssh root\@$vm \"~/$debug_script\" >> $log_file 2>&1");
+        # Save nvidia driver installer logs
+        script_run("rsync root\@$vm:/var/log/nvidia-installer.log ${log_dir}/${vm}_nvidia-installer.log");
+        script_run "mv $log_file $log_dir/${vm}_vgpu_device_status.txt";
+    }
+}
+
+sub post_fail_hook {
+    my $self = shift;
+    diag("Module vgpu post fail hook starts.");
+
+    # Make sure sol console is ok
+    enter_cmd "echo DONE > /dev/$serialdev";
+    reconnect_when_ssh_console_broken unless defined(wait_serial('DONE', timeout => 30));
+
+    my $log_file = "$log_dir/host_vgpu_device_status.txt";
+    print_cmd_output_to_file("virsh list --all", $log_file);
+    print_cmd_output_to_file("nvidia-smi", $log_file);
+    my $cmd = "journalctl --cursor-file /tmp/cursor.txt | grep -r -e 'kernel:' -e gpu -e nvidia | grep -v bash_history";
+    script_run("echo '' >> $log_file");
+    script_run("$cmd >> $log_file");
+    script_run("cp /var/log/nvidia-installer.log $log_dir");
+    save_vgpu_device_status_logs($_) foreach (keys %virt_autotest::common::guests);
+    upload_virt_logs($log_dir, "vgpu");
+    $self->SUPER::post_fail_hook;
+    restore_original_guests() unless get_var('VGPU_SHORT_TEST');
+}
+
+sub test_flags {
+    #continue subsequent test in the case test restored
```

Comment:
Given that only limited machines have gpu, will be good to create separate testsuite for gpu test only, without other feature tests. HDYT?

Reply:
It depends on how many host/guest matrix. If only sle15sp7_guest_on_sle15sp7_host, we have only one test suite need vgpu, it need not create a seprate test suite. If there are more test suites, I'll re-consider this problem.

Reply:
Hard to say now. Nvidia vgpu feature has been supported since sle15sp5, if I remember correctly. So we may *propose* to run on sle15sp5~sle15sp7 hosts. Not sure about final version on 15sp7 LTP.

Reply:
Given there are 3 test suites on sle15sp7:

- sle15sp7_guest_on_sle15sp7_host:      **short** vgpu test on each build + **full** vgpu test on milestone
- sle15sp6_guest_on_sle15sp7_host:      **no** vgpu test on each  build  + **full** vgpu test on milestone
- sle15sp7_guest_on_sle15sp6_host:      **short** vgpu test on each build + **full** vgpu test on milestone

I need re-consider of `ENABLE_VGPU`, `VGPU_SHORT_TEST` and whether creating seperate test suites to support these 3 modes above.

Reply:
My plan:

To create 2 separate vgpu tests and

-  `VGPU_TEST=short/full` to control short test or full test are run, and set it manually for milestone build
- use yaml to schedule test modules
- assign these tests on `amd-zen3-gpu-sut1`
- support with loading vgpu test in mail.pm as well(for flexable test coverage)

The 2 new vgpu tests:
- vgpu_sle15sp6&sp7_guest_on_sle15sp7_host, 
  - for each build: `VGPU_TEST=short`, schedule "install 15sp7 host + vgpu short test"
  - for milestone build: `VGPU_TEST=full`, schedule "install host + install 15sp7 guest + vgpu full test on 15sp7 guest + install 15sp6 guest + vgpu full test on 15sp6 guest"
- vgpu_sle15sp7_guest_on_sle15sp6_host
  - for each build: `VGPU_TEST=short`, schedule "install 15sp6 host + vgpu short test"
  - for milestone build: `VGPU_TEST=full`, schedule "install host + install 15sp7 guest + vgpu full test on 15sp7 guest"

If you think it ok, I'll replace `ENABLE_VGPU=1` with `VGPU_TEST=short/full` in this PR, but create 2 .yaml scheduler files for the two new test suites when sle15sp7 LTP is done.

Reply:
To me, it seems good. I will ask project managers to review and keep you updated.

Reply:
updated.